tests/test_url_handler.py

```diff
@@ -0,0 +1,696 @@
+    results = url_handler.alphabetical_browse(project)
+
+    assert results == expected
+
+
+def test_alphabetical_browse_fake_project():
+    project = 'not a real project'
+    with pytest.raises(http.Http404):
+        url_handler.alphabetical_browse(project)
+
+
+def test_get_metadata():
+    project = factories.ProjectFactory()
+    vals = [x.metadata for x in factories.MetadataValuesFactory.create_batch(3)]
+    metadata = [factories.ProjectMetadataFactory(project=project, metadata=x) for x in vals]
+    expected = [(x, [models.Metadata_Values.objects.get(metadata=x).value]) for x in metadata]
+    results = url_handler.get_metadata(project)
+
+    assert str(sorted(results, key=str)) == str(sorted(expected, key=str))
+
+
+def test_get_metadata_with_valueset():
+    project = factories.ProjectFactory()
+    valset = factories.ValuesetFactory()
+    vals = [factories.ValuesetValuesFactory(valueset=valset).value for i in range(3)]
+    metadata = factories.MetadataFactory(value_sets=[valset])
+    factories.ProjectMetadataFactory(project=project, metadata=metadata)
+    results = url_handler.get_metadata(project)
+
+    for _, value_iter in results:
+        assert list(value_iter).sort() == vals.sort()
+
+
+@pytest.mark.parametrize('posted_data, processed_posted_data, expected', [
+    (
+        {'color': ['blue', 'other_specify']},
+        {'color': ['blue', 'other_specify'], 'color_other': 'magenta'},
+        {'color': ['blue', 'magenta'], 'color_other': 'magenta'}
+    ),
+    (
+        {'color': ['blue', 'other_specify']},
+        {'color': ['blue', 'other_specify']},
+        {'color': ['blue']}
+    ),
+    (
+        {'color': ['blue', 'green']},
+        {'color': ['blue', 'green']},
+        {'color': ['blue', 'green']}
+    ),
+    (
+        {'color': ['other_specify']},
+        {'color': 'other_specify', 'color_other': 'magenta'},
+        {'color': 'magenta', 'color_other': 'magenta'}
+    ),
+    pytest.mark.xfail((
+        {'color': ['other_specify']},
+        {'color': 'other_specify'},
+        {}
+    ))
+])
+def test_handle_metadata(rf, posted_data, processed_posted_data, expected):
+    request = rf.post('/', posted_data)
+    assert url_handler.handle_metadata(request, processed_posted_data) == expected
+
+
+@pytest.mark.parametrize('date_str', [
+    '2006-10-25',
+    '10/25/2006',
+    '10/25/06',
+    'Oct 25 2006',
+    'Oct 25, 2006',
+    '25 Oct 2006',
+    '25 Oct, 2006',
+    'October 25 2006',
+    'October 25, 2006',
+    '25 October 2006',
+    '25 October, 2006'
+])
+def test_validate_date_with_valid_dates(date_str):
+    assert isinstance(url_handler.validate_date(date_str), datetime.date)
+
+
+def test_validate_date_with_invalid_date():
+    date_str = '2006, Oct 25'
+    assert url_handler.validate_date(date_str) is None
+
+
+def test_add_url():
+    project = factories.ProjectWithMetadataFactory()
+    attribute_names = [x.name for x in project.metadata.all()]
```

Comment:
It appears that you only need the name of the first `Project_Metadata`(?)

you can get that with
```python
name = project.metadata.first().name
```
Then you don't have to get the item from a list every time.